Replace deprecated Lwt usages

psafont opened the pull request:
Fixes #13 and also prepares the code for Lwt 5.0

psafont committed:
CP-32138: replace deprecated Lwt_log with Logs_lwt

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

psafont committed:
CP-32138: replace deprecated recv_msg from Lwt_unix

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

lindig reviewed (approved):
This looks good.

psafont commented:
This needs changes in the daemonization of wsproxy (convert to systemd unit?)

lindig commented:
I like that. I would assume that this is not difficult (looking at `wsproxy.ml`). But who spawns wsproxy currently and how would change that?

lindig commented:
`wsproxy` is spawned from `xapi` in `console.ml` using Forkexecd.

psafont commented:
My guess is that socket activation needs to be set up along with the service unit.

mseri commented on cli/wsproxy.ml:
```diff
@@ -45,25 +45,26 @@ let start path handler =
              ~callback:(fun fd ->
-                let iov = Lwt_unix.io_vector ~buffer ~offset:0 ~length:16384 in
-                Lwt_unix.recv_msg ~socket:fd ~io_vectors:[iov])
+                 let io_vectors = Lwt_unix.IO_vectors.create () in
```
🎊

psafont committed:
CP-32138: do not daemonize ourselves

This will be done by whoever spawns us (e.g. systemd)

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

psafont committed:
maintenance: whitespace

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

rwdrich commented on cli/wsproxy.ml:
```diff
@@ -45,25 +45,26 @@ let start path handler =
-                let iov = Lwt_unix.io_vector ~buffer ~offset:0 ~length:16384 in
-                Lwt_unix.recv_msg ~socket:fd ~io_vectors:[iov])
+                 let io_vectors = Lwt_unix.IO_vectors.create () in
+                 Lwt_unix.IO_vectors.append_bytes io_vectors buffer 0 16384;
```
Is it possible to keep the param names?

rwdrich replied:
Apparently not :(

psafont replied:
I tried [:(](https://github.com/ocsigen/lwt/blob/3e5923b40349e2b0b4c0820969dd81cad41b4719/src/unix/lwt_unix.cppo.ml#L692)

gaborigloi commented on cli/wsproxy.ml:
```diff
@@ -13,7 +13,7 @@
 open Wslib
 
-let get_dir_path () = Printf.sprintf "/var/xapi/" 
+let get_dir_path () = Printf.sprintf "/var/xapi/"
```
This could be changed into a variable? :sweat_smile:

psafont replied:
I'm not even sure it's worth using a variable as it is, seeing that it's not configurable and it's only used in a single line.

psafont committed:
CP-32138: simplify path setup

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

gaborigloi commented on cli/wsproxy.ml:
```diff
@@ -117,18 +115,18 @@ let handler sock msg =
 
-let _ = 
+let _ =
+  Logs.set_reporter (Logs_fmt.reporter ());
```
Do we need a different reporter for Lwt?
As suggested by 
https://erratique.ch/software/logs/doc/Logs_lwt/index.html#report_ex and
https://erratique.ch/software/logs/doc/Logs/index.html#setupreporter

psafont replied:
We're debugging to `/dev/null` I don't think it matters if it's synchronous or not in this case.

psafont committed:
CP-32138: Use collaborative log reporter

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

lindig commented:
What is the state of this? Can we merge this?

psafont commented:
The socket/service are not activated when building the packages although they seem to work fine when enabled manually. It needs more changes in the spec files

psafont commented:
Testing didn't go as planned: The connection to the guests are not successful when the wsproxy service is up and running, while managed by systemd.

I'll debug the changes in wsproxy as xapi is correctly opening the socket (wsproxy gets run by systemd)

psafont commented:
Comparison of the log format before and after:

```
Oct 23 12:57:59 localhost wsproxy[26499]: main: About to start connection
Oct 23 12:57:59 localhost wsproxy[26499]: main: Got msg: hybi10:rfb:/var/run/xen/vnc-4
Oct 23 12:57:59 localhost wsproxy[26499]: main: New-style (hybi10) protocol
```
```
Oct 23 13:07:27 localhost wsproxy[17923]: wsproxy: [DEBUG] About to start connection
Oct 23 13:07:27 localhost wsproxy[17923]: wsproxy: [DEBUG] Got msg: hybi10:rfb:/var/run/xen/vnc-8
Oct 23 13:07:27 localhost wsproxy[17923]: wsproxy: [WARNING] Malformed msg: not proxying
```

Processing the message as malformed is not intended, have to look into it. The xapi plumbing looks good so far, however.

psafont committed:
CP-32138: Generate more descriptive messages when match fails

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

psafont committed:
cli: stop logging debug-level messages by default

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

psafont commented:
In order to use socket activation, we cannot create the sockets like we were doing before, we would have to use something like https://github.com/juergenhoetzel/ocaml-systemd

psafont committed:
CP-32138: Do not manage pid files

It's not needed anymore due to systemd

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

psafont committed:
CP-32138: use an already activated socket

Instead of setting up a socket on startup, make the caller
activate the socket and pass it as standard input.

Signed-off-by: Pau Ruiz Safont <pau.safont@citrix.com>

edwintorok commented on cli/wsproxy.ml:
```diff
@@ -13,17 +13,14 @@
+let start handler =
+  let fd_sock_path = "/var/xapi/wsproxy" in
+  Logs_lwt.info (fun m -> m "Starting wsproxy on %s" fd_sock_path) >>= fun () ->
   let fd_sock = Lwt_unix.socket Unix.PF_UNIX Unix.SOCK_STREAM 0 in
```
Can we use Lwt_unix.stdin here?

psafont replied:
yes, haven't had time to push the change ;p
